Listen on fork in EventedFileUpdateChecker

jonathanhefner opened the pull request:
The Listen gem does not notify across process forks, and so forked processes must re-create their listeners.  `EventedFileUpdateChecker` handled this in `updated?` by checking if `Process.pid` was different.  Because file updates may have occurred after the start of the fork but before the call to `updated?`, `updated?` would always return true in this case.

However, the above approach can result in unnecessary application reloading, particularly when using Spring.  That, in turn, can cause unexpected and problematic behavior: #39431, #37591.  This commit changes listener re-creation to occur immediately after fork, and changes `updated?` to return true only when necessary.

---

This is based on top of #40100 and #40133.

This PR should prevent the scenarios described in https://github.com/rails/rails/issues/39431#issuecomment-634931451 and https://github.com/rails/rails/issues/37591#issuecomment-648479317, but it does not address the more general issue of reloading an app from a thread while another thread has acquired a database connection.

jonathanhefner commented on activesupport/lib/active_support/evented_file_update_checker.rb:
```diff
@@ -50,32 +51,20 @@ def initialize(files, dirs = {}, &block)
+      @missing = []
 
       boot!
+      ActiveSupport::ForkTracker.after_fork { boot! }
```
The downside to this is that it keeps a reference to the `EventedFileUpdateChecker` instance in `ActiveSupport::ForkTracker`.  Meaning the instance won't be garbage collected if it is ever abandoned.  I can look into a solution involving `WeakRef` if this is a concern.  Or we could design some kind of `stop!` API.

(It's also why `@dtw` is now recomputed in `boot!`.  Otherwise, if an `EventedFileUpdateChecker` instance is created by a test, and the directories it watches are deleted on teardown, that instance's `boot!` would raise when it's re-triggered by [another test's fork](https://github.com/rails/rails/pull/39718/files#diff-0752a2eca9a80ac701287d0e6666234dR48).)

composerinteralia replied:
https://github.com/rails/rails/blob/a6e98b2267a4470c2de4aa3eb16651c161d98091/actionview/lib/action_view/cache_expiry.rb#L29 looks like one place where we might build a file checker and later abandon it, for what it is worth.

jonathanhefner replied:
That is a good catch! :eyes:  Upon further investigation, I believe that, even without this PR, there is a problem with abandoning checkers due to the way the Listen gem currently works.  I've detailed the problem and proposed some fixes in guard/listen#476, but it hasn't gotten traction yet.  I have a few follow-up Rails PRs in mind, once that issue is addressed.

jonathanhefner replied:
Now that guard/listen#476 has been fixed, I wrote #40100 and #40133 to ensure checkers can be safely abandoned and garbage collected.  I rebased this PR on top of them, and piggy-backed on the finalizer from #40133 to remove the `after_fork` callback from `ActiveSupport::ForkTracker` when the `EventedFileUpdateChecker` instance is garbage collected.

composerinteralia reviewed (commented):
This change seems reasonable to me, and I can confirm it solves https://github.com/rails/rails/issues/37591

jonathanhefner committed:
Stop EventedFileUpdateChecker's listener only

When an `EventedFileUpdateChecker` instance detects a watched directory
that previously did not exist, it calls `shutdown!` and then `boot!` to
reinitialize its underlying listener.  Prior to this commit, `shutdown!`
invoked `Listen.stop` which stops **all** listeners, globally.  This
commit changes `shutdown!` to stop the checker's listener only.

Fixes #38174.

jonathanhefner committed:
Make EventedFileUpdateChecker garbage collectable

The Listen gem dispatches "file changed" events from a separate thread.
This thread holds a reference to the `changed` callback, and runs until
the listener is stopped.  Thus, objects that are part of the `changed`
callback's scope cannot be garbage collected until the listener is
stopped.

This commit isolates the `changed` callback and associated scope in an
`EventedFileUpdateChecker::Core` instance.  This ensures that the
`EventedFileUpdateChecker` instance can be garbage collected.
Additionally, this commit adds a finalizer to the
`EventedFileUpdateChecker` instance, which stops the listener and
ensures that the `EventedFileUpdateChecker::Core` instance can be
garbage collected.

jonathanhefner committed:
Eliminate EventedFileUpdateChecker::PathHelper

jonathanhefner committed:
Listen on fork in EventedFileUpdateChecker

The Listen gem does not notify across process forks, and so forked
processes must re-create their listeners.  `EventedFileUpdateChecker`
handled this in `updated?` by checking if `Process.pid` was different.
Because file updates may have occurred after the start of the fork but
before the call to `updated?`, `updated?` would always return true in
this case.

However, the above approach can result in unnecessary application
reloading, particularly when using Spring.  That, in turn, can cause
unexpected and problematic behavior: #39431, #37591.  This commit
changes listener re-creation to occur immediately after fork, and
changes `updated?` to return true only when necessary.

kaspth commented on activesupport/lib/active_support/evented_file_update_checker.rb:
```diff
@@ -96,15 +67,59 @@ def execute_if_updated
+      def finalizer
+        proc do
+          stop
+          ActiveSupport::ForkTracker.unregister(@after_fork)
```
What about ActiveSupport::ProcessForking as the name here?

jonathanhefner replied:
`ActiveSupport::ForkTracker` was actually introduced by #37312.  I too was pleasantly surprised to find it was already written! :smile:  It hasn't been publicly released yet though, so I can look into renaming it in a separate PR if you would like.

kaspth reviewed (approved):
Also onboard with this! 🙌
